### PR TITLE
Rollups: Skew Mode Batch job Refactor

### PR DESCRIPTION
**Pull request by force2b**
**DRAFT**

# Critical Changes

Complete redesign of the Rollup Batch Job Architecture, primarily for "Skew" Mode. The new structure uses layers of `Abstract` classes that the individual rollup batch jobs Implement. In addition, Skew Mode jobs use a controlling SkewDispatcher batch job that effectively uses PK Chunking to select *small* groups of parent record Id's and then passing that list of Id's into the actual Skew-mode Batch job to roll up. When that completes, control is passed back to the SkewDispatcher to select the new group of parent id's.

Rather than have the nightly schedule execute a single Scheduler class that delegates the appropriate batch job, the 9 individual rollup batch classes all implement `Schedulable` interface to allow the Master Schedule to simply execute each class individually. That also allows for customer admins to manually schedule these jobs as needed.

# Changes

* Changed references for "LDV" to "Skew"
* Unit Test performance/coverage improvements
* Improve Rollup__mdt grouping to improve overall performance
* Refactor the batch job size custom settings fields to reflect the "final" design
* Switch to a RollupType Enum throughout the classes to define which type of job is executing
* For Non-Skew Hard Credit and Contact Soft Credit rollups, implement an Inner Join filter query against the Opportunity/OpporunityContactRole objects to only query parent Accounts or Contacts that have related records. This can be disabled using a custom settings option.
* Added some preliminary support for Recurring Donations rollups

# Issues Closed

# New Metadata

*Classes*
* CRLP_SkewDispatcher_BATCH
* CRLP_Batch_Base
* CRLP_Batch_Base_Skew
* CRLP_Batch_Base_NonSkew
* CRLP_Account_BATCH
* CRLP_Contact_BATCH
* CRLP_AccountSkew_BATCH
* CRLP_ContactSkew_BATCH
* CRLP_AccountSoftCredit_BATCH
* CRLP_ContactSoftCredit_BATCH
* CRLP_AccountSkew_SoftCredit_BATCH
* CRLP_ContactSkew_SoftCredit_BATCH
* CRLP_GAU_BATCH
* CRLP_RollupBatch_SVC

*Custom Settings Fields*
* Rollups_Account_SkewMode_Batch_Size__c
* Rollups_Contact_SkewMode_Batch_Size__c
* Rollups_Account_Soft_Credit_Batch_Size__c
* Rollups_Contact_Soft_Credit_Batch_Size__c
* Rollups_GAU_Batch_Size__c
* Rollups_Limit_on_Attached_Opps_for_Skew__c
* Rollups_Skew_Dispatcher_Batch_Size__c
* Disable_Related_Records_Filter__c

# Deleted Metadata

*Classes*
* CRLP_AbstractRollup_BATCH
* CRLP_Rollup_BATCH
* CRLP_Rollup_LDV_BATCH
* CRLP_Rollup_SoftCredit_LDV_BATCH
* CRLP_RollupScheduler

*Custom Settings Fields*
* Rollups_Account_LDV_Batch_Size__c
* Rollups_Contact_LDV_Batch_Size__c
* Rollups_Allocations_Batch_Size__c
* Rollups_Soft_Credit_Batch_Size__c
* Rollups_Limit_on_Attached_Opps_for_LDV__c

### CODE REVIEW

File: src/classes/ALLO_Rollup_SCHED.cls [hunk 1]
```diff
@@ -42,7 +42,8 @@ public without sharing class ALLO_Rollup_SCHED extends UTIL_AbstractRollup_BATCH
     public void execute(SchedulableContext context) {
         try {
             if (customizableRollupEngineEnabled == false) {
-                ALLO_Rollup_SCHED rollup = new ALLO_Rollup_SCHED(RLLP_OppRollup_UTIL.BATCH_SIZE);
+                ALLO_Rollup_SCHED rollup = new ALLO_Rollup_SCHED(CRLP_RollupBatch_SVC.getBatchSize(
+                        CRLP_RollupProcessingOptions.RollupType.GAU, CRLP_RollupProcessingOptions.BatchJobMode.NonSkewMode));
```
Comment (njjc) [inline src/classes/ALLO_Rollup_SCHED.cls]: Don't GAUs only have "Skew Mode"?

File: src/classes/CRLP_AccountSkew_BATCH.cls [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        Database.executeBatch(new CRLP_SkewDispatcher_BATCH(CRLP_RollupProcessingOptions.RollupType.AccountHardCredit), CRLP_SkewDispatcher_BATCH.BATCH_SIZE);
+    }
+
+    /*******************************************************************************************************
+     * @description Constructs the class - Rollup is for a specific set of Account Id's
+     * @param recordIds Specific set of parent record id's to roll up to
+     * @param lastIdProcessedForChunking If this job is called by the SkewDispatcher, this contains the last Id queried
+     * by the SkewDispatcher Id chunking logic.
+     */
+    public CRLP_AccountSkew_BATCH(List<Id> recordIds, Id lastIdProcessedForChunking) {
```
Comment (njjc) [inline src/classes/CRLP_AccountSkew_BATCH.cls]: Should constructors go first, as far as method ordering?

File: src/classes/CRLP_Batch_Base.cls [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    protected CRLP_RollupProcessingOptions.BatchJobMode jobMode = CRLP_RollupProcessingOptions.BatchJobMode.NonSkewMode;
+
+    /** @description The default filter to apply to Rollups type records for the processing */
+    protected CRLP_RollupProcessingOptions.RollupTypeFilter jobFilter = CRLP_RollupProcessingOptions.RollupTypeFilter.All;
+
+    /** @description This is built one time at the start of the batch process and then re-used throughout
+     * the rest of the batch processing logic. This data is passed in the
+     * ProcessingOptions collection to the processor. It's also used to pre-generate the SOQL used to retrieve
+     * summary or detail records rather than build that in every single batch iteration. Basically this collection
+     * is injected into the CRLP_Rollup_SEL cache to significantly simplify and narrow the pr
```
Comment (njjc) [inline src/classes/CRLP_Batch_Base.cls]: narrow the pr...?

File: src/classes/CRLP_Batch_Base.cls [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    /**
+     * @description Map of Opportunity Id's already processed by Account to avoid duplicating a donation amount
+     * to the Account when rolling up soft credits by the Contact.AccountId
+     */
+    protected Map<Id,Set<Id>> statefulMapOfOppIdsByAcct;
+
+    /** @description Simple counters used to track state */
+    protected Integer batchIteration = 0;
+    protected Integer totalRecordsModified = 0;
+    protected Integer totalCommits=0;
```
Comment (njjc) [inline src/classes/CRLP_Batch_Base.cls]: Spaces before and after the = to match the previous two lines ;)

File: src/classes/CRLP_RollupProcessingOptions.cls [hunk 5]
```diff
@@ -38,13 +38,26 @@
...
 
     /**
-     * @description The processing mode for the calling batch operation - normal or LDV mode. In LDV mode,
+     * @description Defines the type of rollups being executed by the specific batch job. This is used with the
+     * RollupFilterSetting enum to select the correct Rollup__mdt records to use for the entire rollup operation.
+     */
+    public Enum RollupType {
+        AccountHardCredit,
+        ContactHardCredit,
+        AccountSoftCredit,
```
Comment (njjc) [inline src/classes/CRLP_RollupProcessingOptions.cls]: I still think we should name this something that reflects that these are actually contact soft credits that we're aggregating. Maybe AccountContactSoftCredit?

File: src/classes/CRLP_Batch_Base.cls [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+            this.keyField = SObjectType.OpportunityContactRole.fields.ContactId.getName();
+
+        } else if (this.jobType == CRLP_RollupProcessingOptions.RollupType.AccountSoftCredit) {
+            this.detailObjectType = OpportunityContactRole.SObjectType;
+            this.keyField = 'Contact.AccountId';
+        }
+    }
+
+    /**
+     * @description Are active Rollup__mdt records for the job type?
```
Comment (njjc) [inline src/classes/CRLP_Batch_Base.cls]: Missing a word.

File: src/classes/ALLO_Rollup_SCHED.cls (hunk 1 above)
Comment (force2b) [reply src/classes/ALLO_Rollup_SCHED.cls]: Yes. Though technically we could have a non-skew mode for GAU's, I didn't think it made sense.

File: src/classes/CRLP_AccountSkew_BATCH.cls (hunk 2 above)
Comment (force2b) [reply src/classes/CRLP_AccountSkew_BATCH.cls]: Yeah probably. I inserted the scheduler execute method above the batch constructor, but that's definitely not ideal.

File: src/classes/CRLP_RollupProcessingOptions.cls (hunk 5 above)
Comment (force2b) [reply src/classes/CRLP_RollupProcessingOptions.cls]: Good idea

**Comment by force2b** [comment]
Thanks. I made the recommended changes and submitted as a new PR on this same branch.

**Pull request merged**